src/CommercetoolsAuth0.ts

```diff
@@ -0,0 +1,274 @@
+   * When this method is called, the Auth0 user may or may not be associated with a customer
+   * in commercetools. This will be determined based on whether or not a value is provided for
+   * {@see PostLoginSyncParams.accountCustomerId}.
+   *
+   * We expect {@see PostLoginSyncParams.accountCustomerId} to only be undefined when the
+   * user has logged in via a third party OAuth provider such as Google. When logging in using
+   * Google, the {@see preRegistrationSync} method would not have been called, and so we need
+   * to create a customer account in commercetools in this method instead.
+   *
+   * If {@see PostLoginSyncParams.accountCustomerId} is a string, then it means the customer
+   * already exists in commercetools. This is probably because they registered by entering their
+   * email and password in the Auth0 sign-up form, which would have then triggered the
+   * pre-registration flow and a call to {@see preRegistrationSync}, resulting in the customer
+   * account being created in commercetools.
+   *
+   * Regardless of which of the above scenarios applies, we end up with the id of a
+   * commercetools account customer. If we were
+   */
+  public async postLoginSync(options: PostLoginSyncParams): Promise<Customer | null> {
+    let customer: Customer | null = null
+    let accountCustomerId: string | undefined = options.accountCustomerId
+    if (!accountCustomerId) {
+      customer = await this.createCustomer(options)
+      accountCustomerId = customer.id
+    }
+    if (options.mergeCart && options.anonymousCustomerId) {
+      await this.mergeCart({
+        storeKey: options.storeKey,
+        anonymousCustomerId: options.anonymousCustomerId,
+        accountCustomerId,
+      })
+    }
+    return customer
+  }
+
+  /**
+   * Pre-registration synchronisation functionality
+   *
+   * This method should be called when the pre-registration action is executed.
+   * We do not have access to the Auth0 user id at this point, and so cannot set
+   * the externalId property on the commercetools customer record at this point.
+   */
+  public async preRegistrationSync(options: PreRegistrationSyncParams): Promise<Customer> {
+    return await this.createCustomer(options)
+  }
+
+  /**
+   * Post-registration synchronisation functionality
+   *
+   * This method should be called when the post-registration action is executed.
+   * It is used to set the `externalId` on the commercetools customer based on the
+   * given Auth0 user id.
+   */
+  public async postRegistrationSync(options: PostRegistrationSyncParams): Promise<void> {
+    let customer: Customer | null = null
+
+    try {
+      customer = await this.client.getCustomerById({
+        storeKey: options.storeKey,
+        id: options.accountCustomerId,
+      })
+    } catch (e) {
+      throw new CommercetoolsAuth0Error('Unable to retrieve customer', CommercetoolsAuth0ErrorCode.UNEXPECTED_ERROR)
```

Comment:
Are we losing the details from the original error here? Should we include the message or details of the original `e` in the output?